Lab1 - Fibonacci function

Jake88 opened the pull request:
Note: I understand storing a map of all the calculated fibonacci numbers is inefficient given the criteria is only to _print_ them, so I have over-engineered here.

Fixes #193

Review of colleague's PR #183

#### Changes proposed in this PR:
- Solution to lab1 - Fibonacci sequence
- Tests for the fib function

Commit:
feat(lab1): Create fibonacci number generator function

Added a function that takes an integer input and prints out the fibonacci sequence with that many iterations. This also includes the ability to print out negafibonacci.

Commit:
test(lab1): Test main function

Test written for the main function of fibonacci, lab1

Commit:
test(lab1): Test fib function

Add a series of tests to ensure the fibonacci function outputs correctly if given both positive and negative integers.

Commit:
refactor(lab1): Refactor fib function to use modulus

Based on a hint from Dan in the lecture today, I have refactored my implementation to use modulus and taken a more 'positive fibonacci first' approach. I have also added comments to help provide clarity on what each section of the code is doing.

willshen8 commented:
Hi Jake, can I do the review your PR once it's passed CI? For now, can you please remove 'Ready for review' label?

Commit:
fix(typo): Fixed a typo in comment

Jake88 commented:
> Hi Jake, can I do the review your PR once it's passed CI? For now, can you please remove 'Ready for review' label?

More than happy for you to review my changes! :) I've fixed up my typo now, so go for gold.

I do think I need to rebase (or something) my typo commit, but I honestly don't know how to go about doing that so I might need to ask an adult for help!

willshen8 commented:
Hi Jake,
All looks good. Few things to consider:

1. Map would occupy more computing resources compare to array or slice, so computing a larger fib number might have efficiency differences?
2. If you compute large fib number, you might get integer overflow from int, so change it to a larger native type. 
3. A good programming practices is seperation of concerns. You could have a function that generate the map and another function does the print. So that these two functions can be resued if the program gets larger.

Jake88 commented:
> Hi Jake,
> All looks good. Few things to consider:
> 
> 1. Map would occupy more computing resources compare to array or slice, so computing a larger fib number might have efficiency differences?
> 2. If you compute large fib number, you might get integer overflow from int, so change it to a larger native type.
> 3. A good programming practices is seperation of concerns. You could have a function that generate the map and another function does the print. So that these two functions can be resued if the program gets larger.

Thanks!
1. You're right here. In hindsight I should have stuck to the criteria and just stored the last two numbers in the sequence, since we don't need to retain or return the full list!
2. I didn't really think about error handling or scaling too high! I should have tried to break it first and then I would have seen these gaps. ;) Good spot.
3. Sure. I think my fib function is definitely more complex than it needs to be. I guess if the program needed to scale I would look at taking reusable pieces and putting them into their own function.

Jake88 committed:
Merge branch 'master' into lab1

anzboi commented:
Please refrain from merging master into your branch until you have the `merge me` label added by an instructor. This way you will have less master merges in your PR.

anzboi commented on 01_fib/Jake88/main.go:
```diff
@@ -0,0 +1,46 @@
+	}
+	negative := false
+
+	// Check if we are performaing negafibonacci or not, and if so remove the sign from n
```
s/performaing/performing

Jake88 replied:
Actioned. Proof read my new comments and made fixes accordingly.

anzboi commented on 01_fib/Jake88/main.go:
```diff
@@ -0,0 +1,46 @@
+
+func fib(n int) {
+	// Our initial map, with the default starting values needed to begin the sequence.
+	fibonacciMap := map[int]int{
```
This map is overkill. You only use two of its value at any given time, and it is never used once the function returns.
I would suggest you store two variables (perhaps `current` and `previous`) initialized to `1` and `0`, then update them in the loop accordingly.
Alternatively, you can store the map as an unexported `var` outside this function to turn it into a global variable,
```
var fibonacciMap = map[int]int{
  -1: 1,
  0: 0,
}
```
If you go down this path, you will find it harder to write the fibonacci function (and may find it easier to write a recursive function that first checks the map, then populates it if it doesn't exist before returning).
I wouldn't recommend this though.

Jake88 replied:
Actioned. Converted to two variables instead.

anzboi commented on 01_fib/Jake88/main.go:
```diff
@@ -0,0 +1,46 @@
+			if i%2 == 0 {
+				current = -current
+			}
+			fibonacciMap[-i] = current
```
The map values are never accessed for negative integers (except `-1` on the first iteration). This is storing numbers that will never be read.

anzboi commented on 01_fib/Jake88/main_test.go:
```diff
@@ -0,0 +1,49 @@
+		{-7, `"1\n-1\n2\n-3\n5\n-8\n13\n"`},
+		{7, `"1\n1\n2\n3\n5\n8\n13\n"`},
+		{-3, `"1\n-1\n2\n"`},
+		{4, `"1\n1\n2\n3\n"`},
```
Can we add a test for `fib(0)`.
I also don't think you need the backticks here (correct me if I am wrong).

Jake88 replied:
Actioned. Added both a test and functionality to handle fib(0).

anzboi reviewed (changes requested):
Great attempt so far, love that you are playing around with various objects in `go` 👍 .
For the purposes of the lab though, I would like you to rethink the use of the map inside the `fib` function. See my comments for more details.
Also please test any edge cases in your labs (`fib(0)` in this case).

Commit:
refactor(fib): Refactor fib function to use two variables instead of a map

Removed the map from fib function and replaced it with two integers to save memory.

Commit:
feat(fib): Add the ability to handle 0 as in input to fib

Supplying an input of 0 to fib will now print f0 (which is 0) in return.

Commit:
feat(fib): Add a check for integer overflow to fib function

Fib function now has a check to see if the input has caused a number overflow error and prints it out accordingly.
Added a few minor comment updates.

Commit:
test(fib): Add tests for fib function

Added a test for fib(0) and a test to check whether an out of bounds / number overflow error has ocurred.

Commit:
Merge branch 'lab1' of https://github.com/Jake88/go-course into lab1

Commit:
refactor(fib): Additional linting fix

Impatient me went and pushed the code before proof reading my comments. Additional commit to ensure they read correctly.

anzboi commented on 01_fib/Jake88/main.go:
```diff
@@ -0,0 +1,53 @@
+func absolute(n int) (int, bool) {
+	var negative bool
+	if n < 0 {
+		n *= -1
```
`n=-n` works just as well.

anzboi commented on 01_fib/Jake88/main.go:
```diff
@@ -0,0 +1,53 @@
+	for i := 0; i < n; i++ {
+		current := prev1 + prev2
+
+		// Protect against integer overflow / out of bounds error.
```
👍

anzboi commented on 01_fib/Jake88/main_test.go:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	// Test out of bounds
+	fib(100)
```
👍 for testing implemented functionality even if not required.
Normally we would write a separate test for this though.

anzboi reviewed (approved):
LGTM. Just one thing to think about for future labs.
When testing different behaviour (out of bounds vs normal), it should go in a different test.

Jake88 committed:
Merge branch 'master' into lab1
